File: #216_Combination_Sum_III/Solution.java
import java.util.*;

public class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> result = new LinkedList();
        if (k == 0 || k > n || 9 * k < n) return result;
        return combinationSum3(1, k, n);
    }

    private List<List<Integer>> combinationSum3(int lo, int level, int sum) {
    	List<List<Integer>> result = new LinkedList();
    	if (level <= 0 || sum <= 0 || 9 * level < sum || lo > 9) return result;
    	for (int i = lo; i < 10; i++) {
    		if (level == 1 && i == sum) {
    			List bottom = new LinkedList();
    			bottom.add(i);
    			result.add(bottom);
    			return result;
    		} 
    		List<List<Integer>> list = combinationSum3(i + 1, level - 1, sum - i);
    		for (List<Integer> ls : list) {
    			ls.add(0, i);
    			result.add(ls);
    		}
    	}
    	return result;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	List<List<Integer>> result = sol.combinationSum3(3, 9);
    	for (List<Integer> ls : result) {
    		for (Integer it : ls) {
    			System.out.print(it + " ");
    		}
    		System.out.println();
    	}
    }
}

File: #216_Combination_Sum_III/Solution.py
class Solution(object):

	def __init__(self):
		self.re = None

	def combinationSum3(self, k, n):
		"""
		:type k: int number of elements
		:type n: int target sum
		:rtype: List[List[int]]
		"""
		self.re = []
		if k <= 0 or k >= 10 or n < 1 or n > 45:
			return self.re
		nums = [x for x in range(1, 10)]
		self.combine(nums, 0, [], 0, k, n)
		return self.re

	def combine(self, nums, lv, tmp, sum, k, n):
		if lv == k and sum == n:
			self.re.append(list(tmp))
			return
		if not nums:
			return
		if lv < k and sum < n:
			for x in range(0, len(nums) - (k - lv - 1)):
				tmp.append(nums[x])
				self.combine(nums[x + 1:], lv + 1, tmp, sum + nums[x], k, n)
				tmp.pop()

sol = Solution()
print sol.combinationSum3(3, 15)